[MSE][GStreamer] block data flow until source pad is exposed

eocanha opened the pull request:
#### 04cc4a4e2fc235ab43cc38dbe880d9ddb6b6e5dd
<pre>
[MSE][GStreamer] block data flow until source pad is exposed
<a href="https://bugs.webkit.org/show_bug.cgi?id=276251">https://bugs.webkit.org/show_bug.cgi?id=276251</a>

Reviewed by Xabier Rodriguez-Calvar.

In WebKitMediaSrc, there could be a race between main thread (linking
typefind element, see gsturisourcebin.c:setup_typefind) and streaming
thread (pushing data). webKitMediaSrcLoop does wait for source pad to
linked. However, there is no guarantee that peer pad is already active
at that point. This may lead to data loss while peer the pad is being
activated.

This change blocks downstream flow during pad addition, allowing
urisourcebin to link/activate typefind sink pad.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1355">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1355</a>

Original patch by Eugene Mutavchi &lt;Ievgen_Mutavchi@comcast.com&gt;.

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcEmitStreams): Add blocking probes to the Streams Pads, which will be removed after the pad has been added to the element.

Canonical link: <a href="https://commits.webkit.org/280813@main">https://commits.webkit.org/280813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61cf47432b2e006b98288788a4a9002da175eaee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57627 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36955 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61249 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8072 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59755 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44591 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8260 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46660 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5731 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27527 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31447 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7086 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7076 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53404 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7359 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62929 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1541 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7447 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53921 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49807 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54031 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12769 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1312 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32784 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33870 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34954 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -343,6 +344,16 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<Media
 
     for (const RefPtr<Stream>& stream: source->priv->streams.values()) {
+        // Block data flow until pad is exposed.
+        gulong blockId = gst_pad_add_probe(
+            GST_PAD(stream->pad.get()), GST_PAD_PROBE_TYPE_BLOCK_DOWNSTREAM,
+            [](GstPad*, GstPadProbeInfo*, gpointer) -> GstPadProbeReturn {
```
This indentation is weird.

eocanha replied:
It's consistent with the "function with long parameters" example mentioned in the [WebKit coding style guidelines](https://webkit.org/code-style-guidelines/#braces-one-line). The gst_pad_add_probe() function has long parameters, which are placed in next lines and indented one tab with respect to the _add_probe() function.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
```diff
@@ -343,6 +344,16 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<Media
+            }, nullptr, nullptr);
+        auto blockCleanup = WTF::makeScopeExit([&] {
+            gst_pad_remove_probe(GST_PAD(stream->pad.get()), blockId);
+        });
```
Documentation about `GST_PAD_PROBE_REMOVE` is incomplete but I would check if it passes the data and if it does, you don't need the scoped exit.

eocanha replied:
Manually removing the probe is the suggested way to remove a **blocking** probe:
https://gstreamer.freedesktop.org/documentation/additional/design/probes.html?gi-language=c#blocking-probes

There's also the possibility of returning GST_PAD_PROBE_REMOVE, but I'm not sure if the data would be dropped or returned in that case, so better stick to `gst_pad_remove_probe()` in this case. The makeScopeExit ensures that the probe is eventually removed, either if data has ever traversed the pad or not. That's the main difference: gst_pad_remove_probe() would work even if there's no data, and that's an advantage that justifies keeping this code block as is. Do you agree?

calvaris replied:
Well:

> The blocking probe is removed with gst_pad_remove_probe() or when the probe callback return GST_PAD_PROBE_REMOVE. In both cases, and if this was the last blocking probe on the pad, the pad is unblocked and dataflow can continue.

calvaris replied:
I'd also recommend to remove the scope exit and add it to the end of the scope just after line 369 because the code is not error prone with early returns to forget this.

calvaris replied:
And also, even when the probe is blocking, to serve it's purpose, you should check inside it if pad is active and wait there until it is. Otherwise the blocking probe will block, execute and unblock, even with the pad not being active yet, right?

calvaris replied:
It looks like I a mistaken on the blocking stuff, it should work but still, the scope exit is not needed.

calvaris reviewed (approved):
It looks like I a mistaken on the blocking stuff, it should work but still, the scope exit is not needed.

eocanha committed:
[MSE][GStreamer] block data flow until source pad is exposed
https://bugs.webkit.org/show_bug.cgi?id=276251

Reviewed by Xabier Rodriguez-Calvar.

In WebKitMediaSrc, there could be a race between main thread (linking
typefind element, see gsturisourcebin.c:setup_typefind) and streaming
thread (pushing data). webKitMediaSrcLoop does wait for source pad to
linked. However, there is no guarantee that peer pad is already active
at that point. This may lead to data loss while peer the pad is being
activated.

This change blocks downstream flow during pad addition, allowing
urisourcebin to link/activate typefind sink pad.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1355

Original patch by Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>.

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcEmitStreams): Add blocking probes to the Streams Pads, which will be removed after the pad has been added to the element.

Canonical link: https://commits.webkit.org/280813@main
